File: bindings/clojure/src/rockbox/smoke.clj
(ns rockbox.smoke
  "End-to-end smoke test for the Clojure bindings.

  Run: mise exec -- clojure -M:smoke"
  (:require [rockbox.ffi :as ffi]
            [rockbox.metadata :as metadata]
            [rockbox.dsp :as dsp]
            [rockbox.player :as player])
  (:import [java.io File]))

(defn- fail [msg]
  (binding [*out* *err*] (println (str "SMOKE FAILED: " msg)))
  (System/exit 1))

(defn- repo-root ^File []
  (loop [dir (.getAbsoluteFile (File. (System/getProperty "user.dir")))]
    (cond
      (nil? dir) (throw (ex-info "could not find repo root (crates/rocksky/fixtures)" {}))
      (.isDirectory (File. dir "crates/rocksky/fixtures")) dir
      :else (recur (.getParentFile dir)))))

(defn -main [& _]
  (let [fixture (.getPath (File. (repo-root) "crates/rocksky/fixtures/08 - Internet Money - Speak(Explicit).m4a"))]
    (println "ABI version:" (ffi/abi-version))

    ;; -- metadata -------------------------------------------------------
    (try
      (let [meta (metadata/read fixture)
            codec (:codec meta "")]
        (println (str "codec=" codec " title=" (:title meta) " artist=" (:artist meta)
                      " duration_ms=" (:duration_ms meta) " sample_rate=" (:sample_rate meta)))
        (when (empty? codec) (fail "codec label should not be empty"))
        (let [probe (metadata/probe "song.flac")]
          (println "probe('song.flac') =" probe)
          (when-not (= probe "FLAC") (fail (str "expected FLAC, got " probe)))))
      (catch Exception e (fail (str "metadata: " e))))

    ;; -- DSP: -6.0206 dB track gain should HALVE amplitude --------------
    (try
      (let [rate 44100]
        (dsp/with-dsp [d rate]
          (dsp/set-replaygain d :track false 0.0)
          (dsp/set-replaygain-gains d :track-gain-db -6.0206) ; x0.5
          (let [sine (dsp/sine-stereo 1000.0 1.0 rate 16000.0)
                out (dsp/process d sine)
                peak (reduce (fn [m s] (max m (Math/abs (int s)))) 0 out)]
            (println (str "DSP peak after -6.02 dB track gain: " peak " (expected ~8000)"))
            (when-not (<= 7600 peak 8400) (fail (str "peak " peak " out of range"))))))
      (catch Exception e (fail (str "dsp: " e))))

    ;; -- Player (construct only, no audible playback) -------------------
    (try
      (player/with-player [p {:volume 0.0}]
        (let [sr (player/sample-rate p)]
          (println "Player sample_rate=" sr)
          (when (zero? sr) (fail "sample_rate should be > 0"))
          (player/set-volume p 0.0)
          (player/set-queue p [fixture])
          (Thread/sleep 100) ; queue command applied asynchronously
          (let [st (player/status p)
                state (:state st "")
                queue-len (:queue_len st -1)]
            (println (str "Player status: state=" state " queue_len=" queue-len))
            (when-not (= state "stopped") (fail (str "expected stopped, got " state)))
            (when-not (= queue-len 1) (fail (str "expected queue_len 1, got " queue-len))))))
      (catch Exception e (fail (str "player: " e))))

    (println "\nALL CHECKS PASSED ✔")
    (System/exit 0)))

File: bindings/kotlin/src/main/kotlin/org/rockbox/ffi/Smoke.kt
package org.rockbox.ffi

import kotlin.math.abs
import kotlin.system.exitProcess

// End-to-end smoke test for the Kotlin bindings.
//
// Run: mise exec -- gradle smoke   (or ./gradlew smoke)

private fun fail(msg: String): Nothing {
    System.err.println("SMOKE FAILED: $msg")
    exitProcess(1)
}

fun main() {
    val fixture = Fixtures.sample
    println("ABI version: ${abiVersion()}")

    // -- metadata -------------------------------------------------------
    try {
        val meta = Metadata.read(fixture)
        val codec = meta["codec"] as? String ?: ""
        println(
            "codec=$codec title=${meta["title"]} artist=${meta["artist"]} " +
                "duration_ms=${meta["duration_ms"]} sample_rate=${meta["sample_rate"]}",
        )
        if (codec.isEmpty()) fail("codec label should not be empty")

        val probe = Metadata.probe("song.flac")
        println("probe('song.flac') = $probe")
        if (probe != "FLAC") fail("expected FLAC, got $probe")
    } catch (e: Exception) {
        fail("metadata: $e")
    }

    // -- DSP: -6.0206 dB track gain should HALVE amplitude --------------
    try {
        val rate = 44_100L
        Dsp(rate).use { dsp ->
            dsp.setReplaygain(DspReplayGainMode.TRACK, noclip = false, preampDb = 0.0f)
            dsp.setReplaygainGains(trackGainDb = -6.0206f) // x0.5

            val sine = sineStereo(freqHz = 1000.0, seconds = 1.0, rate = rate, amplitude = 16_000.0)
            val out = dsp.process(sine)
            val peak = out.maxOfOrNull { abs(it.toInt()) } ?: 0
            println("DSP peak after -6.02 dB track gain: $peak (expected ~8000)")
            if (peak !in 7_600..8_400) fail("peak $peak out of range")
        }
    } catch (e: Exception) {
        fail("dsp: $e")
    }

    // -- Player (construct only, no audible playback) -------------------
    try {
        val config = Player.Config().apply { volume = 0.0f }
        Player(config).use { player ->
            val sr = player.sampleRate
            println("Player sample_rate=$sr")
            if (sr == 0L) fail("sample_rate should be > 0")

            player.setVolume(0.0f)
            player.setQueue(listOf(fixture))
            Thread.sleep(100) // queue command applied asynchronously

            val st = player.status()
            val state = st["state"] as? String ?: ""
            val queueLen = (st["queue_len"] as? Number)?.toInt() ?: -1
            println("Player status: state=$state queue_len=$queueLen")
            if (state != "stopped") fail("expected stopped, got $state")
            if (queueLen != 1) fail("expected queue_len 1, got $queueLen")
        }
    } catch (e: Exception) {
        fail("player: $e")
    }

    println("\nALL CHECKS PASSED ✔")
}
